nuxeo-core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/coremodel/SQLDocumentProxy.java

```diff
@@ -217,9 +217,60 @@ public void setSystemProp(String name, Serializable value) {
         return target.getSystemProp(name, type);
     }
 
+    public static final String CHANGE_TOKEN_PROXY_SEP = "/";
+
+    /*
+     * The change token for a proxy must reflect the fact that either the proxy (name, parent, acls, etc.) or its target
+     * may be changed.
+     */
     @Override
     public String getChangeToken() {
-        return target.getChangeToken(); // TODO take into account proxy changes as well
+        String proxyToken = proxy.getChangeToken();
+        String targetToken = target.getChangeToken();
+        return getProxyChangeToken(proxyToken, targetToken);
+    }
+
+    protected static String getProxyChangeToken(String proxyToken, String targetToken) {
+        if (proxyToken == null && targetToken == null) {
+            return null;
+        } else {
+            if (proxyToken == null) {
+                proxyToken = "";
+            } else if (targetToken == null) {
+                targetToken = "";
+            }
+            return proxyToken + CHANGE_TOKEN_PROXY_SEP + targetToken;
+        }
+    }
+
+    @Override
+    public boolean validateUserVisibleChangeToken(String userVisibleChangeToken) {
+        if (userVisibleChangeToken == null) {
+            return true;
+        }
+        String[] parts = userVisibleChangeToken.split(CHANGE_TOKEN_PROXY_SEP, 2);
+        if (parts.length != 2) {
+            // invalid format
+            return false;
+        }
+        String proxyToken = parts[0];
+        if (proxyToken.isEmpty()) {
+            proxyToken = null;
+        }
+        String targetToken = parts[1];
+        if (targetToken.isEmpty()) {
+            targetToken = null;
+        }
+        if (proxyToken == null && targetToken == null) {
+            return true;
+        }
+        return proxy.validateUserVisibleChangeToken(proxyToken) && target.validateUserVisibleChangeToken(targetToken);
```

Comment:
`validateUserVisibleChangeToken` are more or less the same for DBS and VCS, can't we do anything for that ?